### PR TITLE
[Merged by Bors] - feat(Analysis/SpecialFunctions/Complex/Log): Exponential tsum is tprod

### PR DESCRIPTION
**Pull request by CBirkbeck**
Taking cexp of an infinite sum of logs gives a tprod.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 1]
```diff
@@ -243,3 +243,23 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
   continuous_iff_continuousAt.2 fun x => h₁.continuousAt.clog (h₂ x)
 
 end LogDeriv
+
+section tsum_tprod
+
+variable {α  ι: Type*}
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:249]:
```suggestion
variable {α ι : Type*}
```

**Review by j-loreaux — COMMENTED** [review]
It feels like we would want much more API around this. For instance:

1. a version for `ℝ`
2. `Summable`-ity conditions
3. `HasSum`/`HasProd` versions

Do you want to add these, or explain why we shouldn't?

**Comment by CBirkbeck** [comment]
> It feels like we would want much more API around this. For instance:
> 
>     1. a version for `ℝ`
> 
>     2. `Summable`-ity conditions
> 
>     3. `HasSum`/`HasProd` versions
> 
> 
> Do you want to add these, or explain why we shouldn't?

Good point, I've added the real version and the HasSum/HasProd versions. What do you mean about the summable conditions? I'm breaking up a big PR which will have things like 
```
lemma summable_multipliable (f : ℕ → ℂ) (hf : Summable f) (hff : ∀ n : ℕ, 1 + f n  ≠ 0) :
    Multipliable (fun n : ℕ => (1 + f n)) := by
```
Is this what you mean? sorry I'm probably just confused.

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 2]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+lemma Real.rexp_tsum_eq_tprod (f : ι → α → ℝ) (hfn : ∀ x n, 0 < f n x)
+    (hf : ∀ x : α, Summable fun n => log ((f n x))) :
+      (rexp ∘ (fun a : α => (∑' n : ι, log ((f n a))))) =
+        (fun a : α => ∏' n : ι, ((f n a))) := by
+  ext a
+  apply (HasProd.tprod_eq ?_).symm
+  apply Real.HasSum_rexp_HasProd f hfn fun a => (hf a).hasSum
+
+lemma Real.summable_cexp_multipliable (f : ι → α → ℝ) (hfn : ∀ x n, 0 < f n x)
+    (hf : ∀ x : α, Summable fun n => log ((f n x))) (a : α):
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:272]:
```suggestion
    (hf : ∀ x : α, Summable fun n => log (f n x)) (a : α):
```

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 3]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+      Multipliable fun b ↦ f b a := by
+  have := (Real.HasSum_rexp_HasProd f hfn fun a => (hf a).hasSum) a
+  use ((rexp ∘ fun a ↦ ∑' (n : ι), log (f n a)) a)
+
+open Complex
+
+lemma Complex.HasSum_cexp_HasProd (f : ι → α → ℂ) (hfn : ∀ x n, f n x ≠ 0)
+    (hf : ∀ x : α, HasSum (fun n => log ((f n x))) (∑' i, log (f i x))) (a : α) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:280]:
```suggestion
    (hf : ∀ x : α, HasSum (fun n => log (f n x)) (∑' i, log (f i x))) (a : α) :
```

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 4]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+       HasProd (fun b ↦ f b a) ((cexp ∘ fun a ↦ ∑' (n : ι), log (f n a)) a) := by
+  apply ((hf a).cexp).congr
+  intro _
+  congr
+  exact funext fun x ↦ exp_log (hfn a x)
+
+lemma Complex.summable_cexp_multipliable (f : ι → α → ℂ) (hfn : ∀ x n, f n x ≠ 0)
+    (hf : ∀ x : α, Summable fun n => log ((f n x))) (a : α):
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:288]:
```suggestion
    (hf : ∀ x : α, Summable fun n => log (f n x)) (a : α):
```

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 5]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+      Multipliable fun b ↦ f b a := by
+  have := (Complex.HasSum_cexp_HasProd f hfn fun a => (hf a).hasSum) a
+  use ((cexp ∘ fun a ↦ ∑' (n : ι), log (f n a)) a)
+
+/--The exponential of a infinite sum of comples logs (which converges absolutely) is an infinite
+product.-/
+lemma Complex.cexp_tsum_eq_tprod (f : ι → α → ℂ) (hfn : ∀ x n, f n x ≠ 0)
+    (hf : ∀ x : α, Summable fun n => log ((f n x))) :
+      (cexp ∘ (fun a : α => (∑' n : ι, log ((f n a))))) =
+        (fun a : α => ∏' n : ι, ((f n a))) := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:298]:
```suggestion
    (hf : ∀ x : α, Summable fun n => log (f n x)) :
      (cexp ∘ (fun a : α => (∑' n : ι, log (f n a)))) =
        (fun a : α => ∏' n : ι, (f n a)) := by
```

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 6]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+
+section tsum_tprod
+
+variable {α ι: Type*}
+
+open Real
+
+lemma Real.HasSum_rexp_HasProd  (f : ι → α → ℝ) (hfn : ∀ x n, 0 < f n x)
+    (hf : ∀ x : α, HasSum (fun n => log ((f n x))) (∑' i, log (f i x))) (a : α) :
+       HasProd (fun b ↦ f b a) ((rexp ∘ fun a ↦ ∑' (n : ι), log (f n a)) a) := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:255]: I can see why you wrote the `HasProd` as `rexp ∘ ...` but I think it makes more sense here to go straight to `Π`, no? I know the proof of some of these results may get *slightly* more complicated as a result, but I think it's worth it for a better statement. Same for the `Complex` version below.
```suggestion
    (hf : ∀ x : α, HasSum (fun n => log (f n x)) (∑' i, log (f i x))) (a : α) :
       HasProd (fun b ↦ f b a) ((rexp ∘ fun a ↦ ∑' (n : ι), log (f n a)) a) := by
```

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean [hunk 7]
```diff
@@ -243,3 +243,61 @@ theorem _root_.Continuous.clog {f : α → ℂ} (h₁ : Continuous f)
...
+  intro _
+  congr
+  exact funext fun x ↦ exp_log (hfn a x)
+
+/--The exponential of a infinite sum of real logs (which converges absolutely) is an infinite
+product.-/
+lemma Real.rexp_tsum_eq_tprod (f : ι → α → ℝ) (hfn : ∀ x n, 0 < f n x)
+    (hf : ∀ x : α, Summable fun n => log ((f n x))) :
+      (rexp ∘ (fun a : α => (∑' n : ι, log ((f n a))))) =
+        (fun a : α => ∏' n : ι, ((f n a))) := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Complex/Log.lean:266]:
```suggestion
lemma Real.rexp_tsum_eq_tprod (f : ι → α → ℝ) (hfn : ∀ x n, 0 < f n x)
    (hf : ∀ x : α, Summable fun n => log (f n x)) :
      (rexp ∘ (fun a : α => (∑' n : ι, log (f n a)))) =
        (fun a : α => ∏' n : ι, (f n a)) := by
```

**Review by j-loreaux — COMMENTED** [review]
Yes, this was essentially the idea I had in mind. Primarily just clean-up below, although it could be nice if two of the statements changed (see below).

File: Mathlib/Analysis/SpecialFunctions/Complex/Log.lean (hunk 6 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/SpecialFunctions/Complex/Log.lean]: Do you mean just prove: `HasProd (fun b ↦ f b a) ∏' n : ι, ((f n a))`?

Comment (j-loreaux) [reply Mathlib/Analysis/SpecialFunctions/Complex/Log.lean]: Yep

Comment (CBirkbeck) [reply Mathlib/Analysis/SpecialFunctions/Complex/Log.lean]: Ok cool, I don't know if I did it the most efficent way, but its not that much worse than what we had before.

**Comment by j-loreaux** [comment]
Nice! Thanks!

bors merge

**Pull request closed without merging**